### ISSUE TITLE
MPI builds fail with parallel HDF5 1.14.0

### ISSUE THREAD

**Issue by quantumsteve** [QMCPACK/qmcpack#4447]
**Describe the bug**

```
-- Configuring done
CMake Error at /home/svh/HDF5/1.14.0/cmake/hdf5-targets.cmake:68 (set_target_properties):
  The link interface of target "hdf5-shared" contains:

    MPI::MPI_C

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /home/svh/HDF5/1.14.0/cmake/hdf5-config.cmake:145 (include)
  /usr/share/cmake-3.25/Modules/FindHDF5.cmake:502 (find_package)
  CMakeLists.txt:646 (find_package)
```

**To Reproduce**
1. release version or git commit hash being built

d0522999905df4cf8f3aadd4e4b18312a2f54236

3. cmake command

CC=mpicc CXX=mpicxx cmake -G Ninja -DHDF5_ROOT=/home/svh/HDF5/1.14.0/ ../

5. full program/test invocation command

(CMake configure step never finished)

7. additional steps

hdf5 1.14.0 built from source with ` -DHDF5_ENABLE_PARALLEL=ON` 

**Expected behavior**

CMake to configure and build without errors.

**System:**
 - system name [e.g. fusiont5, summit]
 - modules loaded [e.g. output of `module list`]
 - other systems where this is reproducible [e.g. "my laptop", "none"]
   my laptop, Ubuntu 20.04

**Additional context**

A similar issue was already reported to HDF5
https://github.com/HDFGroup/hdf5/issues/2404

**Comment by quantumsteve** [comment]
There is a fix in the last commit to develop, but I encountered another error when using that branch.
https://github.com/HDFGroup/hdf5/pull/2400

```
-- Unable to determine HDF5 C flags from HDF5 wrapper.
-- Could NOT find HDF5 (missing: HDF5_LIBRARIES) (found suitable version "1.15.0", minimum required is "1.14")
CMake Error at CMakeLists.txt:698 (message):
  HDF5 not found.  Set HDF5_ROOT
```
I tracked it down to h5cc calling `cc` instead of `mpicc`
```
$ diff /home/svh/HDF5/1.14.0/bin/h5cc /home/svh/HDF5/1.15.0/bin/h5cc
47c47
<       /usr/bin/mpicc $@ `pkg-config --define-variable=prefix=$dir --cflags --libs hdf5`
---
>       /usr/bin/cc $@ `pkg-config --define-variable=prefix=$dir --cflags --libs hdf5`
```
but haven't yet determined why `_PKG_CONFIG_COMPILER` is being defined differently in each case
https://github.com/HDFGroup/hdf5/blob/509fe962d42b3f02d356d2668b04712df06daf7f/src/CMakeLists.txt#L1389

**Comment by quantumsteve** [comment]
With `-DHDF5_ENABLE_PARALLEL=ON` the current `develop` branch is found but isn't picked up as parallel :confused:
```
-- Found HDF5: /home/svh/HDF5/1.15.0/lib/libhdf5.so;/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so (found suitable version "1.15.0", minimum required is "1.14") found components: C 
-- HDF5 IS PARALLEL: FALSE
-- Serial HDF5 library found
-- Using HDF5 non-scalable serial I/O code paths
CMake Warning at CMakeLists.txt:678 (message):
  MPI builds may have performance loss by not using parallel HDF5! (Safe to
  ignore for workstation builds).
```

**Comment by prckent** [comment]
Good to know but hopefully you can stick with 1.12.x for development and work this in parallel?

**Comment by ye-luo** [comment]
1.12.2 is also bad but 1.12.1 works fine.
https://github.com/spack/spack/issues/35893#issuecomment-1456895355

**Comment by ye-luo** [comment]
1.14.1-2 and 1.14.2 remain broken.

**Comment by prckent** [comment]
Thanks for exploring Ye. For both serial and parallel builds? Kitware or HDFGroup or QMCPACK or spack problem? e.g. Do we have to update any API usage? I am happy to explore here since (obviously) if HDF does not work it is game over for users. Perhaps we should update our error messages if we can detect the bad software versions....

**Comment by quantumsteve** [comment]
Fixed in #4708
